pkg/lightning/backend/localhelper.go

```diff
@@ -123,54 +129,109 @@ func (local *local) SplitAndScatterRegionByRanges(ctx context.Context, ranges []
 		} else {
 			splitKeyMap = getSplitKeysByRanges(ranges, regions)
 		}
-		for regionID, keys := range splitKeyMap {
-			var newRegions []*split.RegionInfo
-			region := regionMap[regionID]
-			sort.Slice(keys, func(i, j int) bool {
-				return bytes.Compare(keys[i], keys[j]) < 0
-			})
-			splitRegion := region
-			for j := 0; j < (len(keys)+maxBatchSplitKeys-1)/maxBatchSplitKeys; j++ {
-				start := j * maxBatchSplitKeys
-				end := utils.MinInt((j+1)*maxBatchSplitKeys, len(keys))
-				splitRegionStart := codec.EncodeBytes([]byte{}, keys[start])
-				splitRegionEnd := codec.EncodeBytes([]byte{}, keys[end-1])
-				if bytes.Compare(splitRegionStart, splitRegion.Region.StartKey) < 0 || !beforeEnd(splitRegionEnd, splitRegion.Region.EndKey) {
-					log.L().Fatal("no valid key in region",
-						log.ZapRedactBinary("startKey", splitRegionStart), log.ZapRedactBinary("endKey", splitRegionEnd),
-						log.ZapRedactBinary("regionStart", splitRegion.Region.StartKey), log.ZapRedactBinary("regionEnd", splitRegion.Region.EndKey),
-						log.ZapRedactReflect("region", splitRegion))
-				}
-				splitRegion, newRegions, err = local.BatchSplitRegions(ctx, splitRegion, keys[start:end])
-				if err != nil {
-					if strings.Contains(err.Error(), "no valid key") {
-						for _, key := range keys {
-							log.L().Warn("no valid key",
-								log.ZapRedactBinary("startKey", region.Region.StartKey),
-								log.ZapRedactBinary("endKey", region.Region.EndKey),
-								log.ZapRedactBinary("key", codec.EncodeBytes([]byte{}, key)))
-						}
-						return errors.Trace(err)
-					}
-					log.L().Warn("split regions", log.ShortError(err), zap.Int("retry time", j+1),
-						zap.Uint64("region_id", regionID))
-					retryKeys = append(retryKeys, keys[start:]...)
-					break
-				} else {
-					log.L().Info("batch split region", zap.Uint64("region_id", splitRegion.Region.Id),
-						zap.Int("keys", end-start), zap.Binary("firstKey", keys[start]),
-						zap.Binary("end", keys[end-1]))
-					sort.Slice(newRegions, func(i, j int) bool {
-						return bytes.Compare(newRegions[i].Region.StartKey, newRegions[j].Region.StartKey) < 0
+
+		type splitInfo struct {
+			region *split.RegionInfo
+			keys   [][]byte
+		}
+
+		var syncLock sync.Mutex
+		// TODO, make this size configurable
+		size := utils.MinInt(len(splitKeyMap), runtime.GOMAXPROCS(0))
+		ch := make(chan *splitInfo, size)
+		eg, splitCtx := errgroup.WithContext(ctx)
+
+		for i := 0; i < size; i++ {
```

Comment:
please use another variable name, this loop is too long and is shadowing the outer `i` on line 71 (maybe rename the `i` in line 71 too).

```suggestion
		for splitWorker := 0; splitWorker < size; splitWorker++ {
```